app/ante/fee_checker.go

```diff
@@ -4,51 +4,85 @@ import (
+// as well as a global minimum threshold and computes the tx priority based on the gas price.
+func CheckTxFeeWithMinGasPrices(ctx sdk.Context, tx sdk.Tx, paramKeeper params.Keeper) (sdk.Coins, int64, error) {
 	feeTx, ok := tx.(sdk.FeeTx)
 	if !ok {
 		return nil, 0, errors.Wrap(sdkerror.ErrTxDecode, "Tx must be a FeeTx")
 	}
 
 	fee := feeTx.GetFee().AmountOf(appconsts.BondDenom)
 	gas := feeTx.GetGas()
-	appVersion := ctx.BlockHeader().Version.App
 
-	// global minimum fee only applies to app versions greater than one
-	if appVersion > v1.Version {
-		globalMinGasPrice := appconsts.GlobalMinGasPrice(appVersion)
+	// Ensure that the provided fee meets a minimum threshold for the validator.
+	// This is only for local mempool purposes, and thus
+	// is only ran on check tx.
+	if ctx.IsCheckTx() {
+		defaultMinGasPriceDec, err := sdk.NewDecFromStr(fmt.Sprintf("%f", appconsts.DefaultMinGasPrice))
+		if err != nil {
+			return nil, 0, errors.Wrapf(err, "invalid defaultMinGasPrice: %f", defaultMinGasPriceDec)
+		}
 
-		// convert the global minimum gas price to a big.Int
-		globalMinGasPriceInt, err := sdk.NewDecFromStr(fmt.Sprintf("%f", globalMinGasPrice))
+		err = verifyMinFee(fee, gas, defaultMinGasPriceDec, "insufficient validator minimum fee")
 		if err != nil {
-			return nil, 0, errors.Wrap(err, "invalid GlobalMinGasPrice")
+			return nil, 0, err
 		}
+	}
 
-		gasInt := sdk.NewIntFromUint64(gas)
-		minFee := globalMinGasPriceInt.MulInt(gasInt).RoundInt()
+	// Ensure that the provided fee meets a global minimum threshold.
+	// Global minimum fee only applies to app versions greater than one
+	if ctx.BlockHeader().Version.App > v1.Version {
+		subspace, exists := paramKeeper.GetSubspace(minfee.ModuleName)
+		if !exists {
+			return nil, 0, errors.Wrap(sdkerror.ErrInvalidRequest, "minfee is not a registered subspace")
+		}
 
-		if !fee.GTE(minFee) {
-			return nil, 0, errors.Wrapf(sdkerror.ErrInsufficientFee, "insufficient fees; got: %s required: %s", fee, minFee)
+		if !subspace.Has(ctx, minfee.KeyGlobalMinGasPrice) {
+			return nil, 0, errors.Wrap(sdkerror.ErrKeyNotFound, "GlobalMinGasPrice")
+		}
+
+		var globalMinGasPrice sdk.Dec
+		// Gets the global minimum gas price from the param store
+		// panics if not configured properly
+		subspace.Get(ctx, minfee.KeyGlobalMinGasPrice, &globalMinGasPrice)
+
+		err := verifyMinFee(fee, gas, globalMinGasPrice, "insufficient global minimum fee")
+		if err != nil {
+			return nil, 0, err
 		}
 	}
 
 	priority := getTxPriority(feeTx.GetFee(), int64(gas))
 	return feeTx.GetFee(), priority, nil
 }
 
+// verifyMinFee validates that the provided transaction fee is sufficient given the provided minimum gas price.
+func verifyMinFee(fee math.Int, gas uint64, minGasPrice sdk.Dec, errMsg string) error {
+	// Determine the required fee by multiplying required minimum gas
+	// price by the gas limit, where fee = minGasPrice * gas.
+	minFee := minGasPrice.MulInt(sdk.NewIntFromUint64(gas)).RoundInt()
```

Comment:
[question] should this take the ceiling instead of round? If it rounds down then a transaction could get included that pays a fee slightly less than the one derived used global min gas price.